go/tools/gopackagesdriver/aspect.bzl

```diff
@@ -48,31 +48,29 @@ def file_path(f):
         prefix = "__BAZEL_OUTPUT_BASE__"
     return paths.join(prefix, f.path)
 
-def _go_archive_to_pkg(archive):
-    return struct(
-        ID = str(archive.data.label),
-        PkgPath = archive.data.importpath,
-        ExportFile = file_path(archive.data.export_file),
-        GoFiles = [
-            file_path(src)
-            for src in archive.data.orig_srcs
-            if src.path.endswith(".go")
-        ],
-        CompiledGoFiles = [
-            file_path(src)
-            for src in archive.data.srcs
-            if src.path.endswith(".go")
-        ],
-        OtherFiles = [
-            file_path(src)
-            for src in archive.data.orig_srcs
-            if not src.path.endswith(".go")
-        ],
-    )
-
-def make_pkg_json(ctx, name, pkg_info):
+def _go_archive_to_pkg_json(ctx, name, archive):
     pkg_json_file = ctx.actions.declare_file(name + ".pkg.json")
-    ctx.actions.write(pkg_json_file, content = pkg_info.to_json())
+
+    # Build the args in a config file.
+    args = ctx.actions.args()
+    args.add("--id", archive.data.label)
+    args.add("--pkg-path", archive.data.importpath)
+    args.add("--export-file", archive.data.export_file)
+
+    # Will expand into GoFiles and OtherFiles.
+    args.add_all("--orig-srcs", archive.data.orig_srcs, map_each=file_path)
+
+    # Expands in CompiledGoFiles
+    args.add_all("--data-srcs", archive.data.srcs, map_each=file_path)
+    args.add("--output-file", pkg_json_file)
+    args.use_param_file("@%s.args.txt", use_always = True)
```

Comment:
I could see the param file adding overhead here (may be negligible though). Is `use_always = True` necessary here or could we let bazel decide?

Reply:
Yes, my bad, should have left this out of the commit.